src/multiplex/multiplex.rs

```diff
@@ -530,3 +727,187 @@ impl<S, T, E> Future for Multiplex<S, T>
+        }
+    }
+
+    fn is_outbound(&self) -> bool {
+        !self.is_inbound()
+    }
+
+    fn is_dispatched(&self) -> bool {
+        match self.request {
+            Request::Out(Some(_)) => false,
+            _ => true,
+        }
+    }
+
+    /// Returns true if the exchange is complete
+    fn is_complete(&self) -> bool {
+        // The exchange is completed if the response has been seen and bodies
+        // in both directions are fully flushed
+        self.responded && self.out_body.is_none() && self.in_body.is_none()
+    }
+
+    /// Takes the buffered out request out of the value and returns it
+    fn take_buffered_out_request(&mut self) -> Option<Message<T::Out, Body<T::BodyOut, T::Error>>> {
+        match self.request {
+            Request::Out(ref mut request) => request.take(),
+            _ => None,
+        }
+    }
+
+    fn send_out_chunk(&mut self, chunk: Result<Option<T::BodyOut>, T::Error>) {
+        // Reverse Result & Option
+        let chunk = match chunk {
+            Ok(Some(v)) => Some(Ok(v)),
+            Ok(None) => None,
+            Err(e) => Some(Err(e)),
+        };
+
+        // Get a reference to the sender
+        {
+            let sender = match self.out_body {
+                Some(ref mut v) => v,
+                _ =>  {
+                    return;
+                }
+            };
+
+            if self.out_is_ready {
+                trace!("   --> send chunk; end-of-stream={:?}", chunk.is_none());
+
+                // If there is a chunk (vs. None which represents end of
+                // stream)
+                if let Some(chunk) = chunk {
+                    // Send the chunk
+                    sender.send(chunk);
+
+                    // See if the sender is ready again
+                    match sender.poll_ready() {
+                        Ok(Async::Ready(_)) => {
+                            trace!("   --> ready for more");
+                            // The sender is ready for another message
+                            return;
+                        }
+                        Ok(Async::NotReady) => {
+                            // The sender is not ready for another message
+                            self.out_is_ready = false;
+                            return;
+                        }
+                        Err(_) => {
+                            // The sender is complete, it should be removed
+                        }
+                    }
+                }
+
+                assert!(self.out_deque.is_empty());
+            } else {
+                trace!("   --> queueing chunk");
+
+                self.out_deque.push(chunk);
+                return;
+            }
+        }
+
+        self.out_is_ready = false;
+        self.out_body = None;
+    }
+
+    fn try_poll_in_body(&mut self) -> Poll<Option<T::BodyIn>, T::Error> {
+        match self.in_body {
+            Some(ref mut b) => b.poll(),
+            _ => Ok(Async::NotReady),
```

Comment:
I personally at least prefer to never leave `_` cases if possible, perhaps `None` could be matched explicitly here?